File: design-patterns/ts/abstract-factory/transport/four-wheel-transport-factory.ts
// four-wheel-transport-factory.ts

import AbstractTransportFactory from "./abstract-transport-factory";
import Car from "./car";
import Transport from "./transport";
import Truck from "./truck";

class FourWheelTransportFactory implements AbstractTransportFactory {
    getTransport(type: string): (Transport | null) {
        if (type.toLowerCase() === "car") {
            return new Car();
        }

        if (type.toLowerCase() === "truck") {
            return new Truck();
        }

        return null;
    }
}

export default FourWheelTransportFactory;

File: design-patterns/go/abstract_factory/transport/four_wheel_transport_factory.go
// four_wheel_transport_factory.go

package main

import (
	"errors"
	"strings"
)

type FourWheelTransportFactory struct {
}

func NewFourWheelTransportFactory() (fourWheelTransportFactory *FourWheelTransportFactory) {
	fourWheelTransportFactory = &FourWheelTransportFactory{}
	return
}

func (fourWheelTransportFactory *FourWheelTransportFactory) GetTransport(identifier string) (Transport, error) {
	identifier = strings.ToLower(identifier)

	switch identifier {
	case "car":
		return NewCar(), nil
	case "truck":
		return NewTruck(), nil
	default:
		return nil, errors.New("identifier does not match with any defined four wheel transport")
	}
}
